File: testdata/fixtures/ocaml/CWE-78.ml
(* CWE-78: OS Command Injection – user-controlled data passed to Sys.command
 * or Unix.system allows arbitrary OS command execution. Safe code uses
 * hard-coded literals which are excluded by the rule via not_at.
 *)

let vuln user_input =
  (* ruleid: ocamllint-exec *)
  let _ = Sys.command user_input in
  (* ruleid: ocamllint-exec *)
  let _ = Unix.system user_input in
  (* ruleid: ocamllint-exec *)
  let _ = Sys.command (user_input ^ " arg") in
  (* ruleid: ocamllint-exec *)
  let cmd = user_input in let _ = Unix.system cmd in
  ()

let safe () =
  (* ok: ocamllint-exec *)
  let _ = Sys.command "ls -la" in
  (* ok: ocamllint-exec *)
  let _ = Unix.system "echo hi" in
  (* ok: ocamllint-exec *)
  let _ = Sys.command "fixed --version" in
  ()

File: testdata/fixtures/rust/CWE-78.rs
// CWE-78: OS command injection via Command::new with untrusted input
use std::process::Command;

fn vuln() {
    let input = std::env::var("USER_INPUT").unwrap();
    // ruleid: rust-command-injection
    let _ = Command::new(input).spawn();
    let cmd = std::env::args().next().unwrap();
    // ruleid: rust-command-injection
    let _ = std::process::Command::new(cmd).output();
    let data = std::env::var("CMD").unwrap();
    let propagated = data;
    // ruleid: rust-command-injection
    let _ = Command::new(propagated).arg("x").spawn();
}

fn safe() {
    // ok: rust-command-injection
    let _ = Command::new("ls").arg("-l").spawn();
    // ok: rust-command-injection
    let _ = std::process::Command::new("fixed").output();
    // ok: rust-command-injection
    let safe = "echo";
    let _ = Command::new(safe).spawn();
}
